### PR TITLE
[Diagnostics] Correctly identify location of requirement failure

### PR DESCRIPTION
**Pull request by xedin**
Previously in situations like:

```swift
protocol P {}

struct S<T: P> {
  var value: T
}

_ = S(value: 42)
```

Diagnostic has reported a problem as related to "reference" to `init`
but the failing generic type requirement belongs to `S`, so a
better diagnostic in such case should mention `generic struct S`.

In order to do that we need to first de-duplicate generic type requirement 
checking since constraint system would record generic parameter multiple
times: 1. While resolving `S` and 2. When determining a type of constructor member.

And afterwards add special logic to `RequirementFailure` to determine where
generic signature came from.

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please smoke test compiler performance

**Review by hborla — APPROVED** [review]

**Comment by xedin** [comment]
@swift-ci please smoke test

File: lib/Sema/ConstraintSystem.cpp [hunk 1]
```diff
@@ -126,6 +126,33 @@ void ConstraintSystem::mergeEquivalenceClasses(TypeVariableType *typeVar1,
   // Merge nodes in the constraint graph.
   CG.mergeNodes(typeVar1, typeVar2);
 
+  // Let's try to de-duplicate any generic requirements
+  // associated with given type variable, if it represents
+  // a generic parameter.
+  if (typeVar1->getImpl().getGenericParameter()) {
```
Comment (DougGregor) [inline lib/Sema/ConstraintSystem.cpp]: Should we be looking at all type variables in the equivalence class of `typeVar1` to see if any of them have a corresponding generic parameter?

File: lib/Sema/ConstraintSystem.cpp [hunk 2]
```diff
@@ -126,6 +126,33 @@ void ConstraintSystem::mergeEquivalenceClasses(TypeVariableType *typeVar1,
...
+                   locator->isLastElement(
+                       ConstraintLocator::ConditionalRequirement);
+          }
+
+          return false;
+        });
+
+    llvm::SmallPtrSet<TypeBase *, 4> existingRequirements;
+    for (auto *req : requirements) {
+      auto constraintTy = req->getSecondType();
```
Comment (DougGregor) [inline lib/Sema/ConstraintSystem.cpp]: Should we defensively check here whether the constraint itself has a second type? I feel like the locator checked above could be insufficient.

**Review by DougGregor — APPROVED** [review]

File: lib/Sema/ConstraintSystem.cpp (hunk 2 above)
Comment (xedin) [reply lib/Sema/ConstraintSystem.cpp]: We could `assert` but requirement is always a binary constraint from one type to another.

File: lib/Sema/ConstraintSystem.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/ConstraintSystem.cpp]: I'm not actually sure what do you mean. I think the API doesn't have the best name here, I'll add `isGenericParameter` to make it clearer that this type variable is supposed to represent a particular generic parameter.

File: lib/Sema/ConstraintSystem.cpp (hunk 2 above)
Comment (DougGregor) [reply lib/Sema/ConstraintSystem.cpp]: Okay.

File: lib/Sema/ConstraintSystem.cpp (hunk 1 above)
Comment (DougGregor) [reply lib/Sema/ConstraintSystem.cpp]: We might have two type variables `$T1` and `$T2` that are always equivalent to each other, but only one of them has an associated generic parameter. If `typeVar1` ends up being `$T1` but `$T2` is the generic parameter, we would skip this code. Is that the right behavior?

Comment (xedin) [reply lib/Sema/ConstraintSystem.cpp]: Interesting thought! As I understand current behavior, we only merge type variables representing generic parameters together as a result of `matchDeepEqualityTypes` so both sides have to be generic parameters. For argument/parameter conversions we wait until at least one side is resolved, same for `BindParam`. This check is not strictly necessary I just wanted to make sure that we don't call `gatherConstraints` after every merge. Maybe I should just do `getRepresentative` if check whether it's a generic parameter.

Comment (xedin) [reply lib/Sema/ConstraintSystem.cpp]: I will also try to make this work cache of "fixed" requirements I have.

Comment (xedin) [reply lib/Sema/ConstraintSystem.cpp]: Looks like de-duplication is not longer indeed with caching of fixed requirements, so this is none issue.

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test

**Pull request merged**